lib/ansible/modules/remote_management/manageiq/manageiq_config.py

```diff
@@ -0,0 +1,211 @@
+        else:
+            dct[k] = merge_dct[k]
+
+
+def get_manageiq_config_value(module, name):
+    """ Gets the current value for the given config option.
+    :param module: AnsibleModule making the call
+    :param name: ManageIQ config option name
+    :return: dict of value of the ManageIQ config option
+    """
+    returncode, out, err = module.run_command([
+        "rails",
+        "r",
+        "puts MiqServer.my_server.get_config(:%s).config.to_json" % (name)
+    ], cwd=module.params['vmdb_path'])
+    if returncode != 0:
+        raise Exception("Error getting existing value for ':%s' config: %s" % (name, err))
+
+    return json.loads(out)
+
+
+def create_expected_value(original_value, changes):
+    """ Merges changes into original value to create a merged expected value.
+    :param original_value: dict original value to merge changes into
+    :param changes: dict of changes to merge into original_value to create the expected value
+    :return: dict of the changes merged into the original_value to create an expected value
+    """
+    expected_value = copy.deepcopy(original_value)
+    dict_merge(expected_value, changes)
+    return expected_value
+
+
+def main():
+    # define the module
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(type='str', required=True),
+            value=dict(type='dict', required=False, default={}),
+            vmdb_path=dict(type='str', required=False, default='/var/www/miq/vmdb'),
+        ),
+        supports_check_mode=True
+    )
+
+    # seed the result dict in the object
+    result = dict(
+        changed=False,
+        name='',
+        original_value={},
+        value={},
+        diff={}
+    )
+
+    # get the original value for the given config option
+    try:
+        original_value = get_manageiq_config_value(module, module.params['name'])
+    except Exception as err:
+        module.fail_json(msg=str(err), **result)
+
+    # update the result
+    result['name'] = module.params['name']
+    result['original_value'] = original_value
+
+    # create updated value dictionary
+    expected_value = create_expected_value(original_value, module.params['value'])
+
+    # enable diff mode
+    result['diff'] = {
+        'before': {
+            module.params['name']: original_value
+        },
+        'after': {
+            module.params['name']: expected_value
+        }
+    }
+
+    # if check_mode then stop here
+    if module.check_mode:
+        result['changed'] = original_value != expected_value
+        module.exit_json(**result)
+
+    # update config if difference
+    # else no-op
+    if original_value != expected_value:
+        (update_value_rc, update_value_out, update_value_err) = module.run_command([
+            "rails",
+            "r",
+            "MiqServer.my_server.set_config(:%s => JSON.parse('%s')); MiqServer.my_server.save!" % (module.params['name'], json.dumps(expected_value))
```

Comment:
I'm surprised `set_config` works with just one subtree. Experimentally, this works, but I'm not sure it's safe — why doesn't it blow away all the other subtrees?
This boils down to `Vmdb::Settings.save!(MiqServer.my_server, name => new_value)`.
https://github.com/ManageIQ/manageiq/blob/d4a91f54341/lib/vmdb/settings.rb#L48-L56
my head hurts thinking how all this diffing/merging adds up :-(  I suspect it's simply that SettingsChange records can't represent deletions?
Anyway, I'd feel better if you passed whole tree (merged with desired changes) instead of just one.
BTW, you can `.deep_merge(...)` on ruby side too, if you prefer.

P.S. I'm not sure `my_server.save!` is necessary (though harmless).  `set_config` calls `Vmdb::Settings.save!` which does the work of writing SettingsChange records; `my_server.save!` is just ActiveRecord to save that MiqServer record, if modified — and testing in rails console shows `set_config` did not modify it.

Reply:
@cben the reason i added `MiqServer.my_server.save!` is that without it the configuration change is asynchronous. That is calling `MiqServer.my_server.set_config` puts the operation in a queue. Previous iterations of this module did a wait loop checking for the change to actually take affect. It was a mess. Adding in the `MiQServer.my_server.save!` from my testing is blocking and thus eliminated the need for the crazy wait loop

Reply:
With some testing using the Vmdb::Settings.save! looks to be blocking, so that is sweet.

Reply:
Sorry, I think I mislead you, you may need to go back to `set_config`.
I couldn't understand where you sync/async difference comes from.
`set_config`, besides calling `Vmdb::Settings.save!`, ends up running `MiqServer.my_server.reload_all_server_settings`.  That was a no-op in my testing because the server was down, I only run `rails c`, and I didn't understand how it can help because `rails c` runs in its own process.
But it *can* help — when you have a running server, it *enqueues* a "reload_settings" call. 
Unfortunately, that's async, I don't see how anything can affect other processes synchronously, except it may succeed by chance... :(

I didn't look at your previous loop, but if that's the only direction, I propose dropping it. I'm not sure module should block until all workers saw the reload.  If modules wrote to DB, and queued reload, IMHO that's good enough.